core/src/Kokkos_Tuners.hpp

```diff
@@ -0,0 +1,473 @@
+  using return_type = std::tuple<ValueType>;
+  static return_type build(const node_type& in, double index) {
+    return std::make_tuple(DimensionValueExtractor<node_type>::get(in, index));
+  }
+};
+
+// At levels above the bottom, we tuple_cat the result of our child on the end
+// of our own tuple
+template <class ValueType, class Subtype, class... Indices>
+struct GetMultidimensionalPoint<ValueHierarchyNode<ValueType, Subtype>, double,
+                                Indices...> {
+  using node_type = ValueHierarchyNode<ValueType, Subtype>;
+  using sub_tuple =
+      typename GetMultidimensionalPoint<Subtype, Indices...>::return_type;
+  using return_type = decltype(std::tuple_cat(
+      std::declval<std::tuple<ValueType>>(), std::declval<sub_tuple>()));
+  static return_type build(const node_type& in, double fraction_to_traverse,
+                           Indices... indices) {
+    size_t index         = in.sub_values.size() * fraction_to_traverse;
+    auto dimension_value = std::make_tuple(
+        DimensionValueExtractor<node_type>::get(in, fraction_to_traverse));
+    return std::tuple_cat(dimension_value,
+                          GetMultidimensionalPoint<Subtype, Indices...>::build(
+                              in.get_sub_value(index), indices...));
+  }
+};
+
+template <typename PointType, class ArrayType, size_t... Is>
+auto get_point_helper(const PointType& in, const ArrayType& indices,
+                      std::index_sequence<Is...>) {
+  using helper = GetMultidimensionalPoint<
+      PointType,
+      decltype(std::get<Is>(std::declval<ArrayType>()).value.double_value)...>;
+  return helper::build(in, std::get<Is>(indices).value.double_value...);
+}
+
+template <typename PointType, typename ArrayType>
+struct GetPoint;
+
+template <typename PointType, size_t X>
+struct GetPoint<PointType,
+                std::array<Kokkos::Tools::Experimental::VariableValue, X>> {
+  using index_set_type =
+      std::array<Kokkos::Tools::Experimental::VariableValue, X>;
+  static auto build(const PointType& in, const index_set_type& indices) {
+    return get_point_helper(in, indices, std::make_index_sequence<X>{});
+  }
+};
+
+template <typename PointType, typename ArrayType>
+auto get_point(const PointType& point, const ArrayType& indices) {
+  return GetPoint<PointType, ArrayType>::build(point, indices);
+}
+
+}  // namespace Impl
+
+template <template <class...> class Container, size_t MaxDimensionSize = 100,
+          class... TemplateArguments>
+class MultidimensionalSparseTuningProblem {
+ public:
+  using ProblemSpaceInput = Container<TemplateArguments...>;
+  static constexpr const int space_dimensionality =
+      Impl::get_space_dimensionality<ProblemSpaceInput>::value;
+  static constexpr const size_t max_space_dimension_size = MaxDimensionSize;
+  static constexpr const double tuning_min               = 0.0;
+  static constexpr const double tuning_max               = 0.999;
```

Comment:
Because `array_start + (array_size * 0.999)` is within array, but `array_start + (array_size * 1.0)` isn't. Game for other answers to the same thing though